From 8cdc0f422edd1fe63e3cf50cbf94b442d573c229 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Wed, 11 Jun 2014 11:53:03 -0400
Subject: [kernel] smp: free related resources when failure occurs in hotplug_cfd()

Message-id: <1402487594-26332-4-git-send-email-prarit@redhat.com>
Patchwork-id: 83645
O-Subject: [RHEL7.1 PATCH BZ 1105192 03/14] kernel/smp.c: free related resources when failure occurs in hotplug_cfd()
Bugzilla: 1105192
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

commit 60c323699bb308404dcb60e8808531e02651578a
Author: Chen Gang <gang.chen@asianux.com>
Date:   Wed Sep 11 14:23:22 2013 -0700

    kernel/smp.c: free related resources when failure occurs in hotplug_cfd()

    When failure occurs in hotplug_cfd(), need release related resources, or
    will cause memory leak.

    Signed-off-by: Chen Gang <gang.chen@asianux.com>
    Acked-by: Wang YanQing <udknight@gmail.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/smp.c b/kernel/smp.c
index 948b3ea..df0e765 100644
--- a/kernel/smp.c
+++ b/kernel/smp.c
@@ -49,10 +49,13 @@ hotplug_cfd(struct notifier_block *nfb, unsigned long action, void *hcpu)
     cpu_to_node(cpu)))
    return notifier_from_errno(-ENOMEM);
   if (!zalloc_cpumask_var_node(&cfd->cpumask_ipi, GFP_KERNEL,
-    cpu_to_node(cpu)))
+    cpu_to_node(cpu))) {
+   free_cpumask_var(cfd->cpumask);
    return notifier_from_errno(-ENOMEM);
+  }
   cfd->csd = alloc_percpu(struct call_single_data);
   if (!cfd->csd) {
+   free_cpumask_var(cfd->cpumask_ipi);
    free_cpumask_var(cfd->cpumask);
    return notifier_from_errno(-ENOMEM);
   }
-- 
1.7.1